File: lib/ruby/sample.rb
=begin
Book:
  authors: [Author]
  title: String
  note?: String
  option:
    hoge?: String
    hara?: Bool
=end

class Book < Base
  attr_accessor :authors, :title, :note, :option

  class Book_0 < Base
    attr_accessor :hoge?, :hara?

    def initialize
      @hoge = nil
      @hara = nil
    end

    def to_json_hash
      hash = {}
      # ...
    end

  end

  def initialize
    @authors = []
    @title = ''
    @note = nil
    @option = nil
  end

  def to_json_hash
    hash = {}
    hash[:authors] = @authors
    hash[:title] = @title
    hash[:note] = @note unless @note == nil
    encode(hash)
  end

  # @param [Hash] hash
  def from_json_hash(hash)
    @authors = hash['authors'].to_a
    @title = hash['title'].to_s
    @note = hash['note'] if hash.include? 'note'
    @option = Book_0.new.from_json hash['option'] if hash.include? 'option'
  end
end

class Author < Base
  attr_accessor :name, :age, :sex

  def initialize
    @name = @age = @sex = nil
  end

  def to_json_hash
    hash = {}
    hash[:name] = @name
    hash[:age] = @age
    hash[:sex] = @sex
    encode(hash)
  end
end

File: lib/swift/sample.swift
// Playground - noun: a place where people can play

import UIKit

/*
Book:
authors: [Author]
title: String
note?: String
option:
  hoge?: String
  hara?: Bool
*/

private func encode(obj: AnyObject?) -> AnyObject {
    switch obj {
    case nil:
        return NSNull()
        
    case let ojmObject as Base:
        return ojmObject.toJsonDictionary()
        
    default:
        return obj!
    }
}

private func decodeOptional(obj: AnyObject?) -> AnyObject? {
    switch obj {
    case let x as NSNull:
        return nil
    
    default:
        return obj
    }
}

class Base {
    func toJsonDictionary() -> NSDictionary {
        return NSDictionary()
    }

    class func fromJsonDictionary(hash: NSDictionary) -> Base? {
    }
}


class Book : Base {
    var authors: [Author] = [Author]()
    var title: String = ""
    var note: String?
    var option: Book_0?
    
    class Book_0 : Base {
        var hoge: String?
        var hara: Bool?
    }
    
    override func toJsonDictionary() -> NSDictionary {
        var hash = NSMutableDictionary()
        hash["authors"] = self.authors.map { x in encode(x) }
        hash["title"] = encode(self.title)
        hash["note"] = encode(self.note)
        hash["option"] = encode(self.option)
        return hash
    }
    
    override class func fromJsonDictionary(hash: NSDictionary) -> Book? {
        val this = Book()
        if let xx = hash["authors"] as? [NSDictionary] {
            for x in xx {
                if let obj = Author.fromJsonDictionary(hash: x) {
                    this.authors.append(obj)
                } else {
                    return nil
                }
            }
        } else {
            return nil
        }
        this.title = hash["title"] as String
        this.note = decodeOptional(hash["note"]) as? String
        this.option = Book_0.fromJsonDictionary(hash["option"] as NSDictionary)
        return this
    }
}



class Author : Base {
}
